### PR TITLE
[ENH] native grid search for `TSCGridSearchCV` and `TSRGridSearchCV`

### LINKED ISSUE #6708: [ENH] TSC and TSR tuning - next steps and outstanding items
Next steps for TSC and TSR grid search estimators. Overall goal should be to uncouple this from the `sklearn` grid search, which requires to make use of mtypes that have "length = number of instances", forcing us to use the deprecated `nested_univ` type for unequal length data, or otherwise `numpy3D`.

Related: https://github.com/sktime/sktime/pull/6118

Suggested items, in suggested order:

* `test_multioutput` fails currently for `TSRGridSearchCV`. Since a similar error occurred for a test to be added by the now closed PR https://github.com/sktime/sktime/pull/6054, I suspect it is coming from part of the broadcasting framework. We should diagnose this, fix it for regressors, and classifiers, and add @Cyril-Meyer's test from #6054.
* once resolved, we should ensure that tags are delegated correctly, e.g., `capability:unequal_length`. The "clean" way would be an equivalent of ` _set_delegated_tags` from `_DelegatedForecaster`, for `_DelegatedClassifier` and `_DelegatedRegressor`. We can test whether this works, for `capability:unequal_length`, by adding a test case with a regressor and/or classifier that can deal only with equal length data. If the tag is delegated correctly, the tests should no longer fail.
* Next, we should write the equivalent of `evaluate` in the `forecasting` module for classification and regression - a minimal benchmarking utility.
* Then, we should look at decoupling from `GridSearchCV` from `sklearn`. For this, we can use the `evaluate` feature from the last stepsm abd the `InstanceSplitter` internally, which should be able to handle all mtypes. We can then set the mtype tags to the most general values, letting all (or almost all, e.g., only well-supported) mtypes pass.

Not part of this sequence of actions, but once that is all done, we can start refactoring across modules, with `forecasting`, and try to bring `skopt` and `optuna` support to the classifiers and regressors.

### LINKED ISSUE #10724: [ENH] in `TSCGridSearchCV` and `TSRGridSearchCV`, replace use of `sklearn` grid search by `sktime` native `evaluate` call
Currently, `TSCGridSearchCV` and `TSRGridSearchCV` use the `sklearn` `GridSearchCV` internally.

`scikit-learn` have changed their API, and this will no longer be possible in future versions due to an incompatibility in the tag system.

Hence the internal logic should be replaced by a use of `model_selection.evaluate` from the `classification` resp `regression` module, similar to how `ForecastingGridSearchCV` is working.

`TSCOptCV` already provides non-`sklearn` grid search - and one pathway would be minimal vendoring, or delegation.

However, I would prefer removing the abstraction layer and exposing only the grid search related functionality, i.e., no `hyperactive` vendoring.

### PR DESCRIPTION
**Pull request by yash-sangwan**
#### Reference Issues/PRs
Fixes #10724. Also fixes #6708

#### What does this implement/fix? Explain your changes.
TSCGridSearchCV and TSRGridSearchCV no longer wrap sklearn GridSearchCV.
Both now run grid search natively through the module's own evaluate, on identical pre-computed folds, with parallelization via utils.parallel.

cv_results_ stays in the sklearn dict format.
Defaults stay the same (accuracy_score / r2_score).
refit still accepts bool, str or callable.

New greater_is_better="auto" uses the same direction heuristic as TSCOptCV (via hyperactive), with True/False to override.
As a side effect, passing a plain metric callable now gives real scores instead of silent NaNs.

#### Does your contribution introduce a new dependency? If yes, which one?
No

#### What should a reviewer concentrate their feedback on?
- Direction handling via greater_is_better="auto"
- cv_results_ format and ranking
- Test coverage for the new engine and both classes

#### Did you add any tests for the change?
Yes. New tests for the scoring helper, the shared engine, TSCGridSearchCV and TSRGridSearchCV.

#### Any other comments?
The two pre-existing TSRGridSearchCV multioutput failures remain excluded via _config.py (see #6708). They are unchanged by this PR.

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG].
##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured dependency isolation.

### CODE REVIEW

File: sktime/utils/sklearn/_scoring.py [hunk 1]
```diff
@@ -0,0 +1,324 @@
+"""Resolution of sklearn scoring specifications for sktime tuners.
+
+Turns the ``scoring`` argument of the sktime tuners into metric callables that
+``model_evaluation.evaluate`` can consume, together with the sign and ranking
+direction needed to select the best parameter candidate.
+"""
+# copyright/attribution hyperactive developers, MIT License
+# _guess_greater_is_better is a port of _guess_sign_of_sklmetric from
```
Comment (fkiraly) [inline sktime/utils/sklearn/_scoring.py]: I would not rename this from hyperactive - this makes later refacatoring easier. I.e., if not used directly keep it an 1:1 copy of the vendored code.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great, thanks! Looks very nice.

Some comments:
* I think the `nested_univ` type should now be removed from the internal usage. This was used for compatibility with sklearn.
    * I think it should be possible to pass all panel types now that `evaluate` also accepts. We should avoid too many back/forth conversions in the data types.
* why are we changing test cases? If you want to extend coverage, I would instead suggest to add test cases.
* can you check whether `n_jobs` needs to be deprecated, given that `backend` and `backend_params` are being added, which also contain `n_jobs` for joblib backend?

File: sktime/classification/model_selection/_tune.py [hunk 2]
```diff
@@ -1,19 +1,514 @@
...
+
+
+def _tuner_fitted_params(tuner):
+    """Get the fitted parameters of a tuner.
+
+    The best parameters, and the fitted parameters of ``best_estimator_``
+    if available, the former taking precedence.
+    """
+    fitted_params = {}
+    try:
```
Comment (fkiraly) [inline sktime/classification/model_selection/_tune.py]: bad style: try/except for the purpose of case distinction, especally if we have alternate means to do this.

Instead check for `refit` (perhaps do not make this a separate method?). If there are no fitted params, `get_fitted_params` will not raise exception but return empty `dict`.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks!

I think this looks largely ok.

However, the deprecation should be handled differently, see here:
https://www.sktime.net/docs/developer-guide/deprecation/

In particular, if users pass `n_jobs` and `pre_dispatch`, then it should still work as before. In such a case, they should be passed to `backend_params` (assuming `joblib` backend) and override potential values in there. A warning should be raised, if the parameters are not `"deprecated"`, using `sktime.utils.warn`. Deprecation notices for the release manager (starting with `todo 1.3.0`) should be also added to the code.

File: sktime/classification/model_selection/_tune.py [hunk 3]
```diff
@@ -1,19 +1,552 @@
...
 
 from sktime.classification._delegate import _DelegatedClassifier
+from sktime.classification.model_evaluation import evaluate
+from sktime.exceptions import NotFittedError
+from sktime.utils.parallel import parallelize
+from sktime.utils.sklearn._scoring import _resolve_scoring
+from sktime.utils.warnings import warn
+
+
+def _check_param_grid(param_grid):
```
Comment (fkiraly) [inline sktime/classification/model_selection/_tune.py]: this is duplicated code, not DRY. Please move to `utils.sklearn`

Comment (fkiraly) [reply sktime/classification/model_selection/_tune.py]: locations: classification, forecasting, regression

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great! Could you kindly explain why we need the `_FixedSplitter`?

File: sktime/classification/model_selection/_tune.py (hunk 3 above)
Comment (yash-sangwan) [reply sktime/classification/model_selection/_tune.py]: Moved _check_param_grid to sktime.utils.sklearn and updated the call sites.

**Comment by yash-sangwan** [comment]
Just opened the upstream PR on `skbase` here: sktime/skbase#586
Once everything looks good, we can merge this one and I'll start prepping the final cleanup for Step 4

**Review by fkiraly — CHANGES_REQUESTED** [review]
This looks really great, but also it suffers from a few AI hallucination problems, and/or has not been checked sufficiently by a human.

* detail in public docstrings is removed, for parameters that currently are well documented. This PR seems to remove explanations without a good reason.
    * a secondary question is whether this maps on differences in behaviour - which, optimally, should not be the case
* fitted attributes are removed without reason, such as `n_features_in_`

Can you kindly look at this?

I would also request to move the common code to `sktime.base._panel`, where code common in regression and classification (and clustering sometimes) is currently placed.

**Review by fkiraly — APPROVED** [review]
Great!

**Review by fkiraly — CHANGES_REQUESTED** [review]
Wait a minute, I think you have not actioned my request regarding the docstrings.

File: sktime/regression/model_selection/_tune.py [hunk 4]
```diff
@@ -1,125 +1,112 @@
...
 
-        Refer :ref:`User Guide <cross_validation>` for the various
-        cross-validation strategies that can be used here.
+        Splits are computed once, before the search, so that all parameter
+        candidates are evaluated on the same folds.
 
-    verbose : int
-        Controls the verbosity: the higher, the more messages.
+    verbose : int, default=0
+        Controls the verbosity. If positive, the number of fits is printed.
```
Comment (fkiraly) [inline sktime/regression/model_selection/_tune.py:100]: here information is missing - what values can this take, etc.

Previously, this was explained.

File: sktime/regression/model_selection/_tune.py [hunk 5]
```diff
@@ -128,13 +115,11 @@ class TSRGridSearchCV(_DelegatedRegressor):
...
-        If ``False``, the ``cv_results_`` attribute will not include training
-        scores.
-        Computing training scores is used to get insights on how different
-        parameter settings impact the overfitting/underfitting trade-off.
-        However computing the scores on the training set can be computationally
-        expensive and is not strictly required to select the parameters that
-        yield the best generalization performance.
+        Whether to include training scores in ``cv_results_``.
+
+        Deprecated, and will be removed in sktime 1.3.0. Train scores are not
```
Comment (fkiraly) [inline sktime/regression/model_selection/_tune.py:120]: Thi was changed. I am also not clear about why these are no longer computed. Based on the algorithm, train scores *are* computed.

File: sktime/regression/model_selection/_tune.py [hunk 6]
```diff
@@ -145,96 +130,111 @@ class TSRGridSearchCV(_DelegatedRegressor):
         Has the same effect as applying ColumnEnsembleRegressor wrapper to self.
         If False, the same best parameter is selected for all variables.
 
+    greater_is_better : "auto", bool, optional, default="auto"
+        Whether higher values of the reported metric are better, used to rank
+        the parameter candidates.
+
+        - "auto" determines the direction from the metric. Scikit-learn scorers
+          are higher-is-better by their sign convention. For metric callables,
```
Comment (fkiraly) [inline sktime/regression/model_selection/_tune.py:138]: I do not think this is correct. MSE is lower-is-better, for instance.

File: sktime/regression/model_selection/_tune.py [hunk 7]
```diff
@@ -145,96 +130,111 @@ class TSRGridSearchCV(_DelegatedRegressor):
...
 
-        For multi-metric evaluation, this is present only if ``refit`` is
-        specified.
-
     best_index_ : int
-        The index (of the ``cv_results_`` arrays) which corresponds to the best
+        The index in the ``cv_results_`` arrays which corresponds to the best
         candidate parameter setting.
 
-        The dict at ``search.cv_results_['params'][search.best_index_]`` gives
```
Comment (fkiraly) [inline sktime/regression/model_selection/_tune.py:184]: why is this information removed?

File: sktime/regression/model_selection/_tune.py [hunk 8]
```diff
@@ -145,96 +130,111 @@ class TSRGridSearchCV(_DelegatedRegressor):
...
-        specified.
-
-        This attribute is not available if ``refit`` is a function.
+        Mean cross-validated score of ``best_estimator_``.
+        Not available if ``refit`` is a callable.
 
     best_params_ : dict
         Parameter setting that gave the best results on the hold out data.
 
-        For multi-metric evaluation, this is present only if ``refit`` is
```
Comment (fkiraly) [inline sktime/regression/model_selection/_tune.py:177]: why is this information removed?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Thanks - I left some comments relate to the docstring.

File: sktime/regression/model_selection/_tune.py (hunk 4 above)
Comment (yash-sangwan) [reply sktime/regression/model_selection/_tune.py:100]: Yes, the old >1 / >2 / >3 levels were sklearn’s and no longer apply, so I’ll document what this implementation actually prints

File: sktime/regression/model_selection/_tune.py (hunk 5 above)
Comment (yash-sangwan) [reply sktime/regression/model_selection/_tune.py:120]: evaluate only writes test scores, it does not return train scores.
To keep train scores we would need a second scoring pass on the training folds, which this PR does not do.

I can make that clearer in the docstring. If you want train scores kept, I can add that as a follow-up.

File: sktime/regression/model_selection/_tune.py (hunk 6 above)
Comment (yash-sangwan) [reply sktime/regression/model_selection/_tune.py:138]: Agreed the wording is easy to misread.

I meant sklearn *scorers* like ``neg_mean_squared_error``, which are higher-is-better by sign.
Raw ``mean_squared_error`` is lower-is-better, and that is already in the next line.

I’ll reword the "auto" bullet so that distinction is clearer.

File: sktime/regression/model_selection/_tune.py (hunk 7 above)
Comment (yash-sangwan) [reply sktime/regression/model_selection/_tune.py:184]: This is still there on the right, just reworded:

The dict at ``cv_results_["params"][best_index_]`` gives the parameter setting for the best model, i.e., is identical with ``best_params_``.

I dropped the "highest mean score" line and the "only if refit is specified" line because best_index_ is always set now. I can put the original wording back if you prefer.

File: sktime/regression/model_selection/_tune.py (hunk 8 above)
Comment (yash-sangwan) [reply sktime/regression/model_selection/_tune.py:177]: Removed because it is no longer true.
On this PR ``best_params_`` is always set, including when ``refit=False``.
I did not keep "present only if refit is specified" because that would document a restriction the code no longer has.

**Review by fkiraly — APPROVED** [review]
Ok, I think this is good for now, we may return to this later.
